### PR TITLE
CBG-352: Create internal structure for cache config

### PR DESCRIPTION
**Pull request by JRascagneres**
Also includes CBG-353

### CODE REVIEW

File: db/database_test.go [hunk 1]
```diff
@@ -388,7 +388,10 @@ func TestGetRemovedAsUser(t *testing.T) {
 	// Manually remove the temporary backup doc from the bucket
 	// Manually flush the rev cache
 	// After expiry from the rev cache and removal of doc backup, try again
-	db.DatabaseContext.revisionCache = NewShardedLRURevisionCache(KDefaultRevisionCacheCapacity, db.DatabaseContext.revCacheLoader, db.DatabaseContext.DbStats.StatsCache())
+	revCacheOptions := RevisionCacheOptions{
+		Size: KDefaultRevisionCacheCapacity,
+	}
+	db.DatabaseContext.revisionCache = NewShardedLRURevisionCache(&revCacheOptions, db.DatabaseContext.revCacheLoader, db.DatabaseContext.DbStats.StatsCache())
```
Comment (bbrks) [inline db/database_test.go]: For these tests I guess we can pass nil in, as `NewShardedLRURevisionCache` will use the default constants if unset.

File: rest/server_context.go [hunk 2]
```diff
@@ -538,13 +538,23 @@ func (sc *ServerContext) _getOrAddDatabaseFromConfig(config *DbConfig, useExisti
...
+
 	var revCacheSize uint32
 	if config.RevCacheSize != nil && *config.RevCacheSize > 0 {
 		revCacheSize = *config.RevCacheSize
 	} else {
 		revCacheSize = db.KDefaultRevisionCacheCapacity
 	}
 
+	var revCacheShardNum uint16
+	revCacheShardNum = 0
```
Comment (bbrks) [inline rest/server_context.go]: I guess this will be updated as part of the external/config options PR?

**Review by bbrks — DISMISSED** [review]
Changes make sense to me - We can do a bit of tidy up in the tests using default values by passing nil.

File: rest/server_context.go (hunk 2 above)
Comment (JRascagneres) [reply rest/server_context.go]: Yes. Currently set to 0 to preserve default value and then will match this to the appropriate config option as part of CBG-355

File: db/database_test.go (hunk 1 above)
Comment (JRascagneres) [reply db/database_test.go]: Very true.

File: db/revision_cache.go [hunk 3]
```diff
@@ -20,16 +20,27 @@ type ShardedRevisionCache struct {
 	numShards uint16
 }
 
+type RevisionCacheOptions struct {
+	Size        *uint32
```
Comment (adamcfraser) [inline db/revision_cache.go]: Can switch to uint32, and move the nil handling to the config processing side.

File: db/revision_cache.go [hunk 4]
```diff
@@ -20,16 +20,27 @@ type ShardedRevisionCache struct {
...
-	if capacity == 0 {
-		capacity = KDefaultRevisionCacheCapacity
+	cacheSize := KDefaultRevisionCacheCapacity
+
+	if cacheOptions != nil && cacheOptions.ShardNumber != 0 {
+		numShards = cacheOptions.ShardNumber
+	}
+
+	if cacheOptions != nil && cacheOptions.Size != nil {
+		cacheSize = *cacheOptions.Size
```
Comment (adamcfraser) [inline db/revision_cache.go]: As discussed above, better to move this to the config processing side.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/channel_cache.go [hunk 5]
```diff
@@ -15,6 +15,7 @@ var (
 	DefaultChannelCacheMinLength = 50               // Keep at least this many entries in cache
 	DefaultChannelCacheMaxLength = 500              // Don't put more than this many entries in cache
 	DefaultChannelCacheAge       = 60 * time.Second // Keep entries at least this long
+	DefaultChannelCacheMaxNumber = 0                // Default of 0 will mean there is no limit
```
Comment (JRascagneres) [inline db/channel_cache.go]: Set default to 50k

**Comment by bbrks** [comment]
One TODO to remove default config handling in `NewLRURevisionCache`

Screenshot 2019-05-03 at 19 20 19

File: db/revision_cache_test.go [hunk 6]
```diff
@@ -8,10 +8,10 @@ import (
 	"strconv"
 	"testing"
 
-	"github.com/couchbase/sync_gateway/base"
 	goassert "github.com/couchbaselabs/go.assert"
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
+
```
Comment (adamcfraser) [inline db/revision_cache_test.go]: whitespace cleanup

**Review by adamcfraser — APPROVED** [review]

File: db/revision_cache_test.go (hunk 6 above)
Comment (JRascagneres) [reply db/revision_cache_test.go]: Done. I ticked a weird import ordering option at some point.

**Pull request merged**